cmd/kanikoExecute.go

```diff
@@ -74,11 +109,51 @@ func runKanikoExecute(config *kanikoExecuteOptions, telemetryData *telemetry.Cus
 				log.SetErrorCategory(log.ErrorConfiguration)
 				return errors.Wrapf(err, "failed to read registry url %v", config.ContainerRegistryURL)
 			}
-			containerImageTag := fmt.Sprintf("%v:%v", config.ContainerImageName, strings.ReplaceAll(config.ContainerImageTag, "+", "-"))
-			dest = []string{"--destination", fmt.Sprintf("%v/%v", containerRegistry, containerImageTag)}
+
 			commonPipelineEnvironment.container.registryURL = config.ContainerRegistryURL
-			commonPipelineEnvironment.container.imageNameTag = containerImageTag
+
+			// Docker image tags don't allow plus signs in tags, thus replacing with dash
+			containerImageTag := strings.ReplaceAll(config.ContainerImageTag, "+", "-")
+
+			if config.ContainerMultiImageBuild {
+				log.Entry().Debugf("Multi-image build activated for image name '%v'", config.ContainerImageName)
+				i, err := docker.ImageListWithFilePath(config.ContainerImageName, config.ContainerMultiImageBuildExcludes, fileUtils)
+				if err != nil {
+					return fmt.Errorf("failed to identify image list for multi image build: %w", err)
+				}
+				if len(i) == 0 {
+					return fmt.Errorf("no docker files to process, please check exclude list")
+				}
+				for image, file := range i {
+					log.Entry().Debugf("Building image '%v' using file '%v'", image, file)
+					containerImageNameAndTag := fmt.Sprintf("%v:%v", image, containerImageTag)
+					dest = []string{"--destination", fmt.Sprintf("%v/%v", containerRegistry, containerImageNameAndTag)}
+					buildOpts := append(config.BuildOptions, dest...)
+					err = runKaniko(file, buildOpts, execRunner)
+					if err != nil {
+						return fmt.Errorf("failed to build image '%v' using '%v': %w", image, file, err)
+					}
+					commonPipelineEnvironment.container.imageNames = append(commonPipelineEnvironment.container.imageNames, image)
+					commonPipelineEnvironment.container.imageNameTags = append(commonPipelineEnvironment.container.imageNameTags, containerImageNameAndTag)
+				}
+
+				// for compatibility reasons also fill single imageNameTag field with "root" image in commonPipelineEnvironment
+				// only consider if it has been built
+				// ToDo: reconsider and possibly remove at a later point
+				if len(i[config.ContainerImageName]) > 0 {
```

Comment:
If only one `Dockerfile` was found, we could take this and be more compatible.

Reply:
in the past only Dockerfile in the root was considered. Thus the current condition should be the most compatible way or do I overlook s.th.?

Reply:
I meant the case `len(i[config.ContainerImageName]) == 0`. In this case `commonPipelineEnvironment.container.imageNameTag` stays empty and the following steps might do nothing if they do not support `config.ContainerMultiImageBuild`, right? Or am I missing something?

Reply:
you are right. It will stay empty as it did before if I don't overlook something.
So far there was a hard link to the `Dockerfile` in the root. If there was none the environment also stayed empty.